python-frontend/src/main/java/org/sonar/plugins/python/api/SonarLintCache.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.python.api;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import org.sonar.api.Beta;
+import org.sonar.api.batch.sensor.cache.ReadCache;
+import org.sonar.api.batch.sensor.cache.WriteCache;
+import org.sonarsource.api.sonarlint.SonarLintSide;
+
+/**
+ * Component used in SonarLint to transfer data in memory between plugins.
+ * At the time of writing, this is used only by the DBD plugin, to consume IRs produced by DBD custom rules in SonarLint context.
+ */
+@SonarLintSide()
+@Beta
+public class SonarLintCache implements ReadCache, WriteCache {
+
+  private final Map<String, byte[]> cache = new HashMap<>();
+
+
+  @Override
+  public InputStream read(String s) {
+    if (!contains(s)) {
+      throw new IllegalArgumentException(String.format("SonarLintCache does not contain key \"%s\"", s));
+    }
+    return new ByteArrayInputStream(cache.get(s));
+  }
+
+  @Override
+  public boolean contains(String s) {
+    return cache.containsKey(s);
+  }
+
+  @Override
+  public void write(String s, InputStream inputStream) {
+    try {
+      write(s, inputStream.readAllBytes());
+    } catch (IOException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Override
+  public void write(String s, byte[] bytes) {
+    if (contains(s)) {
+      throw new IllegalArgumentException(String.format("Same key cannot be written to multiple times (%s)", s));
+    }
+    cache.put(s, bytes);
+  }
+
+  @Override
+  public void copyFromPrevious(String s) {
+    throw new UnsupportedOperationException("SonarLintCache does not allow to copy from previous.");
```

Comment:
I see your point. Maybe we could throw `IllegalArgumentException` if the key doesn't exist and `IllegalStateException` otherwise, but it feels a bit artificial. To me `UnsupportedOperationException` seems more intentional because this operation is not supported by the SonarLint cache, wdyt?

Reply:
No, you are right that UnsupportedOperationException is the best fit for this situation.
Can we create a PR for the plugin API and update the Javadoc for visibility?

(On the other hand, I do realize that I am very nitpicky here, so we could also merge it as is and treat that plugin API PR as optional / do it later)

Reply:
I think that eventually, we would like to have SonarLint itself implementing the Caching APIs (meaning we could drop our `SonarLintCache`) so it might be a bit premature to change the Javadoc at this stage.
I would suggest merging it to the feature branch and we can have a chat about it to decide what's best